skopt/gbrt_opt.py

```diff
@@ -0,0 +1,154 @@
+"""Gradient boosted trees based minimization algorithms."""
+
+import numpy as np
+
+from scipy import stats
+from scipy.optimize import fmin_l_bfgs_b
+from scipy.optimize import OptimizeResult
+
+from sklearn.base import clone
+from sklearn.utils import check_random_state
+
+from .gbt import GradientBoostingQuantileRegressor
+from .utils import extract_bounds
+
+
+def _expected_improvement(X, surrogate, best_y, xi=0.01):
+    """Evaluate expected improvement for `surrogate` model at `x`
+
+    Parameters
+    ----------
+    X : array-like
+        Values at which to evaluate the acquisition function.
+
+    best_y: float, optional
+        The previous best value on which we want to improve.
+
+    xi: float, default 0.01
+        By how much must a new point improve over the previous best
+        value.
+    """
+    X = np.asarray(X)
+    if X.ndim == 1:
+        X = np.expand_dims(X, axis=0)
+
+    # low and high are assumed to be the 16% and 84% quantiles
+    low, mu, high = surrogate.predict(X).T
+    # approximate the std dev, if the pdf is gaussian this is exact
+    std = (high - low) / 2.
+
+    ei = np.zeros(len(mu))
+
+    mask = std > 0
+    improvement = best_y - xi - mu[mask]
+    exploit = improvement * stats.norm.cdf(improvement / std[mask])
+    explore = std[mask] * stats.norm.pdf(improvement / std[mask])
+    ei[mask] = exploit + explore
+
+    ei = -ei # we are being used in a minimizer
+    return ei
+
+
+def _random_point(lower, upper, n_points=1, random_state=None):
```

Comment:
Needed it at several points and was planning ahead for when we have categorical variables as well so I made a function out of it.